Bare `Tuple` should mean `Tuple[Any, ...]`.

gvanrossum opened the pull request:
Fix #2184. (See also python/typing#284.)

JukkaL commented on test-data/unit/check-tuples.test:
```diff
@@ -820,3 +820,20 @@ from typing import List, Tuple
+def f(a: Tuple) -> None: pass
+f(())
+f((1,))
+f(('', ''))
```
Call with a non-tuple argument such as an int.

gvanrossum commented:
Ouch. I just realized that I broke the behavior of `Tuple[()]`. I added a test for that but didn't finish that thought.

gvanrossum commented:
@JukkaL Another look? I added the test with int you requested, and fixed the issue I introduced with `Tuple[()]` -- that was a little awkward since the information about whether the original code was `Tuple` or `Tuple[()]` was lost by the time it reached typeanal.py.

I also looked into the difference with the next block and decided to leave that alone (although reducing it to `self.builtin_type('builtins.tuple', [t.args[0].accept(self)])` made all tests pass, I decided to give it the benefit of the doubt, maybe the line number is important after all).

elazarg commented:
If I understand correctly, this flag may be useful in order to write `Union[]` as a synonym for `UninhabitedType` (i.e. NoReturn) , while still allowing `Union` to be `Union[Any]` (which is actually `Any`).

gvanrossum commented:
Sorry, Union[] (like Tuple[]) is invalid syntax. We use Tuple[()] as an
escape, but Union[()] would be too ugly for such a purpose.

JukkaL commented on test-data/unit/check-tuples.test:
```diff
@@ -820,3 +820,22 @@ from typing import List, Tuple
+from typing import Tuple
+def f(a: Tuple[()]) -> None: pass
+f(())
+f((1,))  # E: Argument 1 to "f" has incompatible type "Tuple[int]"; expected "Tuple[]"
```
Maybe update the output of an empty tuple to be `Tuple[()]` instead of `Tuple[]`?

JukkaL commented:
Other than the comment above, this looks good to me.

gvanrossum commented:
About Tuple[] vs. Tuple[()] in output... I have the patch for that ready,
but I'm not sure I like it. I had to change some other tests that had
Tuple[] in their output, and somehow I have a feeling that for a user,
*seeing* Tuple[] is more indicative of "the type of an empty tuple" than
Tuple[()], which only works for obscure reasons. So I think I'd like to
leave this for a later time.

JukkaL commented:
OK, sounds reasonable `Tuple[()]` really isn't the most intuitive thing ever.
